sympy/matrices/expressions/matexpr.py

```diff
@@ -989,12 +989,20 @@ def transpose(self):
     def matrix_form(self):
         if self.first != 1 and self.higher != 1:
             raise ValueError("higher dimensional array cannot be represented")
-        # Remove one-dimensional identity matrices:
-        # (this is needed by `a.diff(a)` where `a` is a vector)
-        if self.first == Identity(1):
-            return self.second.T
-        if self.second == Identity(1):
-            return self.first
+
+        def _get_shape(elem):
+            if isinstance(elem, MatrixExpr):
+                return elem.shape
+            return (None, None)
+
+        if _get_shape(self.first)[1] != _get_shape(self.second)[1]:
+            # Remove one-dimensional identity matrices:
+            # (this is needed by `a.diff(a)` where `a` is a vector)
+            if _get_shape(self.second) == (1, 1):
+                return self.first*self.second[0, 0]
+            if _get_shape(self.first) == (1, 1):
+                return self.first[1, 1]*self.second.T
+            raise ValueError("incompatible shapes")
```

Comment:
And this. (There is also a ValueError on line 991 that is not tested, although out of this PR.)

Reply:
It should never be reached. If it is, then there's a bug somewhere.

Reply:
With "the one above" below, I refer to this. Either there is a testable case where incorrect input makes it reach this part of the code or one can remove it. The next person will have significantly less ideas how to trigger the Exception compared to you.

Reply:
The tests should make sure that the code produces the correct results. We are not testing every line of the code. This statement does not need testing.